### PR TITLE
Do Not Add Certificate To Root Store

### PR DESCRIPTION
**Pull request by LLLXXXCCC**

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+                            var kind = PropertySetAbstractValueKind.Unflagged;
+
+                            if (constructorMethod.Parameters.Length > 0)
+                            {
+                                var valueContent = argumentValueContentAbstractValues[0].LiteralValues;
+
+                                if (constructorMethod.Parameters[0].Type.Equals(storeNameTypeSymbol) &&
+                                    valueContent.Contains(6) ||
+                                    constructorMethod.Parameters[0].Type.SpecialType == SpecialType.System_String &&
+                                    valueContent.Contains("Root"))
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: >Contains("Root") [](start = 49, length = 16)

Is the API case sensitive?

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+                                if (allResults == null)
+                                {
+                                    return;
+                                }
+
+                                foreach (var kvp in allResults)
+                                {
+                                    if (kvp.Value.Equals(HazardousUsageEvaluationResult.Flagged))
+                                    {
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: Do you want to do anything in the MaybeFlagged case?  I'm fine either way; your call.  Just want to make sure you considered it.

File: src/Microsoft.NetCore.Analyzers/Core/Security/SystemSecurityCryptographyResources.resx [hunk 3]
```diff
@@ -180,6 +180,15 @@
   <data name="DoNotDisableSchUseStrongCryptoMessage" xml:space="preserve">
     <value>{0} disables TLS 1.2 and enables SSLv3</value>
   </data>
+  <data name="DoNotInstallRootCert" xml:space="preserve">
+    <value>Do Not Add Certificate To Root Store</value>
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/SystemSecurityCryptographyResources.resx]: >Certificate [](start = 22, length = 11)

Certificates

File: src/Microsoft.NetCore.Analyzers/Core/Security/SystemSecurityCryptographyResources.resx [hunk 4]
```diff
@@ -180,6 +180,15 @@
...
+  </data>
+  <data name="DoNotInstallRootCertDescription" xml:space="preserve">
+    <value>By default, the Trusted Root Certification Authorities certificate store is configured with a set of public CAs that has met the requirements of the Microsoft Root Certificate Program. Since all trusted root CAs can issue certificates for any domain, an attacker can pick a weak or coercible CA that you install by yourself to target for an attack – and a single vulnerable, malicious or coercible CA undermines the security of the entire system. To make matters worse, these attacks can go unnoticed quite easily.</value>
+  </data>
+  <data name="DoNotInstallRootCertMessage" xml:space="preserve">
+    <value>It is not recommended to install additional certificate into your operating system's Root certificate store/keychain</value>
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/SystemSecurityCryptographyResources.resx]: >It is not recommended to install additional certificate into your operating system's Root certificate store/keychain [](start = 11, length = 116)

Adding certificates to the operating system's trusted root certificates increases the risk of incorrectly authenticating an illegitimate certificate.

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+                            var kind = PropertySetAbstractValueKind.Unflagged;
+
+                            if (constructorMethod.Parameters.Length > 0)
+                            {
+                                var valueContent = argumentValueContentAbstractValues[0].LiteralValues;
+
+                                if (constructorMethod.Parameters[0].Type.Equals(storeNameTypeSymbol) &&
+                                    valueContent.Contains(6) ||
+                                    constructorMethod.Parameters[0].Type.SpecialType == SpecialType.System_String &&
+                                    valueContent.Any(s => s.ToString().ToLower().Equals("root", StringComparison.Ordinal)))
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: >s.ToString().ToLower().Equals("root", StringComparison.Ordinal) [](start = 58, length = 63)

How about String.Equals(s.ToString(), "root", StringComparison.OrdinalIgnoreCase)?  That'll avoid the extra string created from ToLower().

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                            if (constructorMethod.Parameters.Length > 0)
+                            {
+                                var valueContent = argumentValueContentAbstractValues[0].LiteralValues;
+
+                                if (constructorMethod.Parameters[0].Type.Equals(storeNameTypeSymbol) &&
+                                    valueContent.Contains(6) ||
+                                    constructorMethod.Parameters[0].Type.SpecialType == SpecialType.System_String &&
+                                    valueContent.Any(s => string.Equals(s.ToString(), "root", StringComparison.OrdinalIgnoreCase)))
+                                {
+                                    kind = valueContent.Count == 1 ? PropertySetAbstractValueKind.Flagged : PropertySetAbstractValueKind.MaybeFlagged;
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: >valueContent.Count == 1  [](start = 43, length = 24)

If there were three values, 6, "root", and "ROOT", you'd want to treat that as flagged.

**Comment by dotpaul** [comment]
Just handle the case of multiple flagged contructor arguments, and looks good to me.

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs (hunk 6 above)
Comment (LLLXXXCCC) [reply src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: Yes, I think Contains one of these value is good. I won't flag this situation as MaybeFlagged.

**Comment by dotpaul** [comment]
If you want some test cases for the "maybe" rule, I think you should be able to do something like:
```csharp
X509Store s;
if (flag) s = new X509Store("Root");
else s = new X509Store("My");
s.Add(new X509Certificate2());
```

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs (hunk 6 above)
Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: I liked your previous approach of:
- All bad values => Flagged
- Some, but not all, bad value => MaybeFlagged

I'll change other places in the code to do the same.

Comment (LLLXXXCCC) [reply src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: Agree. But the code will be long.

Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/Core/Security/DoNotInstallRootCert.cs]: I'll write a helper method to do this.  Your code is fine for now.

**Pull request merged**